### PR TITLE
Create NodeDidNotRunGC test

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
Rollout Task: [AB#4370](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4370) - while trying to rollout the feature, we found this bug.

The goal of this PR is to create a test that throws a NodeDidNotRunGC error during summarization.

Background:
When rolling out closing the summarizer instead of refreshing v2, which downloaded the snapshot from the server before closing the summarizer, telemetry was indicating there was a lot more refreshes than expected. The ratio was expected to be around 4-6 refresh to 1 close, and it turned out to be a little less than 2 refreshes to 1 close in preview at 25%. This meant we were exacerbating a problem.

After an investigation, we found that `DataObject`s were being created during summarization which put the `summarizerNode` system in a half-baked state. The summary upload would succeed, but the state after summarization would be inconsistent. After upload we would get a `NodeDidNotRunGC`. This test intends to mimic a customer scenario. This does not solve the problem, but we now can verify our solutions.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+export const TestDataObjectType1 = "@fluid-example/test-dataStore1";
+export const TestDataObjectType2 = "@fluid-example/test-dataStore2";
+class TestDataObject2 extends DataObject {
+	public get _root() {
+		return this.root;
+	}
+	public get _context() {
+		return this.context;
+	}
+	// If this datastore created DDSes it would be even more nodes created without running gc
```
Comment (NicholasCouri) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:45]: >// If this datastore created DDSes it would be even more nodes created without running gc [](http://example.com/codeflow?start=1&length=89)

? do we need this comment ?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	public get _context() {
+		return this.context;
+	}
+
+	private readonly datastoreKey = "TestDataObject2";
+
+	protected async hasInitialized() {
+		// This can be fired synchronously as well, which would cause an even worse half state
```
Comment (NicholasCouri) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:60]: >// This can be fired synchronously as well, which would cause an even worse half state [](http://example.com/codeflow?start=2&length=86)

same here - or perhaps better context ?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	[],
+	[],
+);
+const dataStoreFactory1 = new DataObjectFactory(
+	TestDataObjectType1,
+	TestDataObject1,
+	[SharedMap.getFactory(), SharedMatrix.getFactory()],
+	[],
+	[],
+	// it would be nice to move away from this flow
```
Comment (NicholasCouri) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:100]: >// it would be nice to move away from this flow [](http://example.com/codeflow?start=1&length=47)

why ? add more info ?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	mixinSummaryHandler(getComponent),
+);
+
+const innerRequestHandler = async (request: IRequest, runtime: IContainerRuntimeBase) =>
```
Comment (NicholasCouri) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:104]: >innerRequestHandler [](http://example.com/codeflow?start=6&length=19)

Out of curiosity, do we need this ?

**Review by NicholasCouri — APPROVED** [review]
:shipit:

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+				"default",
+			);
+			const newDO = await dataStoreFactory1.createInstance(rootDataObject.containerRuntime);
+			rootDataObject._root.set("store", newDO.handle);
+			const summarizerClient = await createSummarizer(provider, container);
+
+			await provider.ensureSynchronized();
+
+			// This should not fail
+			await assert.rejects(waitForSummary(summarizerClient), "expected NodeDidNotRunGC");
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:178]: Worht validating that the error is NodeDidNotRunGC

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import { IContainerRuntimeBase, IFluidDataStoreFactory } from "@fluidframework/runtime-definitions";
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import { FluidDataStoreRuntime, mixinSummaryHandler } from "@fluidframework/datastore";
+import { pkgVersion } from "../packageVersion";
+
+const runtimeOptions: IContainerRuntimeOptions = {
+	summaryOptions: {
+		summaryConfigOverrides: { state: "disabled" },
+	},
+	gcOptions: { gcAllowed: true },
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:33]: This is not needed anymore.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	rootDataObjectType,
+	RootTestDataObject,
+	[],
+	[],
+	[],
+);
+const dataStoreFactory1 = new DataObjectFactory(
+	TestDataObjectType1,
+	TestDataObject1,
+	[SharedMap.getFactory(), SharedMatrix.getFactory()],
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:97]: This is not needed.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:100]: Yeah, I would rather not have this comment.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+const runtimeOptions: IContainerRuntimeOptions = {
+	summaryOptions: {
+		summaryConfigOverrides: { state: "disabled" },
+	},
+	gcOptions: { gcAllowed: true },
+};
+export const rootDataObjectType = "@fluid-example/rootDataObject";
+export const TestDataObjectType1 = "@fluid-example/test-dataStore1";
+export const TestDataObjectType2 = "@fluid-example/test-dataStore2";
+class TestDataObject2 extends DataObject {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:38]: This is never used. TestDataObject1 creates RootTestDataObject.

**Comment by agarwal-navin** [comment]
This test is very similar to the one I added here - https://github.com/microsoft/FluidFramework/pull/15988.
I will make sure to merge the two if I commit my changes.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	protected async hasInitialized() {
+		// This can be fired synchronously as well, which would cause an even worse half state
+		await this.init();
+	}
+
+	private async init() {
+		const dataObject2 = await rootDataObjectFactory.createInstance(
+			this.context.containerRuntime,
+		);
+		this.root.set(this.datastoreKey, dataObject2.handle);
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:68]: This is not the same scenario as what is seen in tasks component right? This will end up always failing, no matter what we do - For example, even if did a second summary attempt, this will fail. Whereas in case of tasks component, this would succeed.
Maybe adding a check that the data store does not exist, before creating a new one will make it closer to the real one?
Also, wasn't there something about this being not awaited?

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 1 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:45]: Probably not

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 9 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:68]: Ok, I can add a modification where this does succeed eventually.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 4 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:104]: Probably not

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 3 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:100]: OK, I'll remove this

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:60]: I'll remove it

File: packages/test/test-utils/src/TestSummaryUtils.ts [hunk 10]
```diff
@@ -28,11 +28,11 @@ import { timeoutAwait } from "./timeoutUtils";
 
 const summarizerClientType = "summarizer";
 
-async function createSummarizerCore(
+export async function createSummarizerCore(
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/TestSummaryUtils.ts:31]: This is not supposed to be exported. The other createSummarizer APIs should be used instead. If you are adding something generic to the API, you should add it to the other APIs instead.

File: packages/test/test-utils/src/TestSummaryUtils.ts [hunk 11]
```diff
@@ -164,8 +164,12 @@ export async function createSummarizerWithTestConfig(
...
  * submitted summary.
  */
-export async function summarizeNow(summarizer: ISummarizer, reason: string = "end-to-end test") {
-	const result = summarizer.summarizeOnDemand({ reason });
+export async function summarizeNow(
+	summarizer: ISummarizer,
+	reason: string = "end-to-end test",
+	refreshLatestAck?: boolean,
+) {
+	const result = summarizer.summarizeOnDemand({ reason, refreshLatestAck });
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/TestSummaryUtils.ts:172]: What does this refreshLatestAck option do?

File: packages/runtime/container-runtime/package.json [hunk 12]
```diff
@@ -112,6 +112,10 @@
 		"typescript": "~4.5.5"
 	},
 	"typeValidation": {
-		"broken": {}
+		"broken": {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/package.json:115]: From what I have heard, any change that updates this should go in next. There have been unexpected errors because of these changes.

File: packages/runtime/container-runtime/src/summary/summarizer.ts [hunk 13]
```diff
@@ -299,6 +299,10 @@ export class Summarizer extends EventEmitter implements ISummarizer {
 		}
 	}
 
+	public get refreshAndSummarizeLock(): Promise<void> | undefined {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizer.ts:302]: Where is this used?

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/summarizer.ts:302]: It's used in tests. We should not add public methods for tests only. There should be other ways to make the test work.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+				const deferredAck = new Deferred();
+				summarizingContainer.on("op", (op) => {
+					if (op.type === MessageType.SummaryAck) {
+						deferredAck.resolve(op);
+					}
+				});
+				await provider.ensureSynchronized();
+				await deferredAck.promise;
+				await (summarizer as Summarizer).refreshAndSummarizeLock;
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts:191]: Why is this needed?

File: packages/test/test-utils/src/TestSummaryUtils.ts (hunk 11 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-utils/src/TestSummaryUtils.ts:172]: What's happening is that the tinylicious tests are failing because the ack either isn't being processed in time, or it's being processed too late. I'm getting a summarizeOnDemand error.

File: packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts (hunk 14 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/summarizerWithLocalChanges.spec.ts]: Removed, but it was needed for tinylicious tests to complete. The summarizeLock was causing `summarizeOnDemand` to fail

File: packages/runtime/container-runtime/src/summary/summarizer.ts (hunk 13 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/summary/summarizer.ts]: Removed

File: packages/runtime/container-runtime/package.json (hunk 12 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/package.json]: Reverted

File: packages/test/test-utils/src/TestSummaryUtils.ts (hunk 10 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-utils/src/TestSummaryUtils.ts]: Unexported
